writer/writer.go

```diff
@@ -15,27 +15,380 @@
-		return errors.New("unable to convert metadata to JSON")
+func (av *ArtifactsWriter) archiveData(updates *updateBucket) error {
+	destination := filepath.Join(av.updateLocation, "data")
+
+	// create directory and file for archiving data
+	if err := os.MkdirAll(destination, os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+	dataArchive, err := os.Create(filepath.Join(destination, updates.location+".tar.gz"))
+	if err != nil {
+		return err
+	}
+	defer dataArchive.Close()
+
+	// we need to ensure correct ordering of files
+	var dataContent []ReadArchiver
+
+	for _, update := range updates.updateArtifacts {
+		dataContent = append(dataContent,
+			NewFileArchiver(update.path, update.info.Name()))
+	}
+	updates.archivedPath = dataArchive.Name()
+	return av.writeArchive(dataArchive, dataContent, true)
+}
+
+func (av ArtifactsWriter) archiveHeader(updates []os.FileInfo) error {
+	archive, err := os.Create(filepath.Join(av.updateLocation, "header.tar.gz"))
+	if err != nil {
+		return err
+	}
+	defer archive.Close()
+
+	// we need to ensure correct ordering of files
+	var hCnt []ReadArchiver
+	// header-info
+	sr := NewJSONStreamArchiver(av.getHeaderInfo(updates), "header-info")
+	hCnt = append(hCnt, sr)
+
+	for _, update := range updates {
+		bucket, ok := av.updates[update.Name()]
+
+		if !ok {
+			return errors.New("artifacts writer: invalid update bucket")
+		}
+		updateTarLocation := filepath.Join("headers", update.Name())
+
+		// files
+		hCnt = append(hCnt, NewJSONStreamArchiver(bucket.files,
+			filepath.Join(updateTarLocation, "files")))
+		// type-info
+		hCnt = append(hCnt, NewFileArchiver(filepath.Join(bucket.path, "type-info"),
+			filepath.Join(updateTarLocation, "type-info")))
+		// meta-data
+		hCnt = append(hCnt, NewFileArchiver(filepath.Join(bucket.path, "meta-data"),
+			filepath.Join(updateTarLocation, "meta-data")))
+		// checksums
+		for _, upd := range bucket.updateArtifacts {
+			fileName := strings.TrimSuffix(upd.name, filepath.Ext(upd.name)) + ".sha256sum"
+			hCnt = append(hCnt, NewStreamArchiver(upd.checksum,
+				filepath.Join(updateTarLocation, "checksums", fileName)))
+		}
+		// signatures
+		for _, upd := range bucket.updateArtifacts {
+			fileName := strings.TrimSuffix(upd.name, filepath.Ext(upd.name)) + ".sig"
+			fr := NewFileArchiver(filepath.Join(bucket.path, "signatures", fileName),
+				filepath.Join(updateTarLocation, "signatures", fileName))
+			hCnt = append(hCnt, fr)
+		}
+		// TODO: scripts
+		//tarContent = append(tarContent, NewPlainFile(filepath.Join(bucket.path, "scripts"), filepath.Join("headers", update.Name(), "scripts")))
+	}
+	return av.writeArchive(archive, hCnt, true)
+}
+
+func (av *ArtifactsWriter) removeCompressedHeader() error {
+	// remove temporary header file
+	return os.Remove(filepath.Join(av.updateLocation, "header.tar.gz"))
+}
+
+func (av *ArtifactsWriter) createArtifact(files []os.FileInfo) error {
+	artifact, err := os.Create(filepath.Join(av.updateLocation, "artifact.mender"))
```

Comment:
I suppose the name `artifact.mender` should be variable, no? So you can have `beaglebone-update-2016-Q1.mender`, for example.